module/zfs/gzip.c

```diff
@@ -53,16 +54,34 @@ typedef uLongf zlen_t;
 size_t
 gzip_compress(void *s_start, void *d_start, size_t s_len, size_t d_len, int n)
 {
+	int ret;
 	zlen_t dstlen = d_len;
 
 	ASSERT(d_len <= s_len);
 
 	/* check if hardware accelerator can be used */
 	if (qat_dc_use_accel(s_len)) {
-		if (qat_compress(QAT_COMPRESS, s_start,
-		    s_len, d_start, d_len, &dstlen) == CPA_STATUS_SUCCESS)
-			return ((size_t)dstlen);
-		/* if hardware compress fail, do it again with software */
+		/*
+		 * QAT requires some additional scratch space to work with
+		 * to prevent warnings. See qat_compress() for details.
+		 */
+		size_t add_buf_len = d_len;
+		void *add_buf = zio_data_buf_alloc(add_buf_len);
+
+		ret = qat_compress(QAT_COMPRESS, s_start, s_len, d_start,
+		    d_len, add_buf, add_buf_len, &dstlen);
```

Comment:
Since the caller never needs to access this additonal space it's aukward to force it to do the allocation.  It really shouldn't need to worry about this when it can all be handled internally in the qat code.  How about renaming `qat_compress()` to `qat_compress_impl()` and then have `qat_compress()` be a wrapper function which handles allocating the needed memory for QAT_COMPRESS before calling `qat_compress_impl()`.

This would then allow you to move the more involved status checking about success/failure/compressed size all in to the wrapper function.  Callers `qat_compress()` wouldn't then need to deal with it at all and could using the existing interface.